sentinelhub/download/session.py

```diff
@@ -126,3 +134,125 @@ def _fetch_token(self, request: DownloadRequest) -> JsonDict:
+    other Python processes during multiprocessing parallelization.
+
+    How to use it:
+
+    .. code-block:: python
+
+        thread = SessionSharingThread(session)
+        thread.start()
+
+        # Run a parallelization process here
+        # Use collect_shared_session() to retrieve the session with other processes
+
+        thread.stop()
+    """
+
+    _EXTRA_MEMORY_BYTES = 100
+
+    def __init__(
+        self, session: SentinelHubSession, *args: Any, memory_name: str = _DEFAULT_SESSION_MEMORY_NAME, **kwargs: Any
+    ):
+        """
+        :param session: A Sentinel Hub session to be used for sharing its authentication token.
+        :param args: Positional arguments to be propagated to `threading.Thread` parent class.
+        :param memory_name: A unique name for the requested shared memory block.
+        :param kwargs: Keyword arguments to be propagated to `threading.Thread` parent class.
+        """
+        super().__init__(*args, **kwargs)
+
+        self.session = session
+        self.memory_name = memory_name
+
+        if self.session.refresh_before_expiry is None:
+            raise ValueError(f"Given instance of {self.session.__class__.__name__} must be self-refreshing")
+        self._refresh_time = self.session.refresh_before_expiry
+
+        self._stop_event = Event()
+        self._is_memory_shared_event = Event()
+
+    def start(self) -> None:
+        """Start running the thread.
+
+        After starting the thread it also waits for the token to be shared. This way no other process would try to
+        access the memory before it even exists."""
+        super().start()
+        self._is_memory_shared_event.wait()
+
+    def run(self) -> None:
+        """A running thread is running an infinite loop of sharing a token and waiting for token to expire. The loop
+        ends only when the thread is stopped."""
+        self._stop_event.clear()
+
+        while not self._stop_event.is_set():
+            token = self.session.token
+            self._share_token(token)
+
+            sleep_until_refresh_time = token["expires_at"] - time.time() - self._refresh_time
+            if sleep_until_refresh_time > 0:
+                self._stop_event.wait(timeout=sleep_until_refresh_time)
+
+    def _share_token(self, token: JsonDict) -> None:
+        """A token is encoded into bytes and written into a shared memory block.
+
+        Note that the `SharedMemory` object allocates extra `self._EXTRA_MEMORY_BYTES` bytes of memory because the
+        length of encoded token can vary a bit.
+        """
+        encoded_token = json.dumps(token).encode()
+
+        if self._is_memory_shared_event.is_set():
+            memory = SharedMemory(name=self.memory_name)
+        else:
+            memory = SharedMemory(
+                create=True,
+                size=len(encoded_token) + self._EXTRA_MEMORY_BYTES,
```

Comment:
this one worries me a bit because it could break if we significantly change the token, but i'm not sure there is another way of doing it

Reply:
In practice token size changes maybe +-3 characters because `"expires_at"` can have different number of decimals. I put `100` characters spare space just in case. Unless they change something at SH service that it would be returning tokens of very varying sizes, the current extra memory should be enough.

Reply:
alright, lets leave it then